### PR TITLE
Tag unused warning as such

### PR DESCRIPTION
**Pull request by serras**
This patch uses the new "unnecessary" or "deprecated" tags for diagnostics to show unused variables/functions/... dimmed in the editor, as discussed in https://github.com/haskell/ghcide/issues/165#issuecomment-692532054

### CODE REVIEW

**Comment by serras** [comment]
Mandatory screenshot (notice the grey color)

Captura de pantalla 2020-09-16 a las 20 45 06

**Comment by ndmitchell** [comment]
Awesome! One question - I usually don't turn off unused variables, just unused top-level bindings. It would be great to still get the grey color. Could we always turn these warnings on, but suppress them from making it to the diagnostics if we turned them on rather than the user?

**Comment by serras** [comment]
That sounds like a good idea, it shouldn't be more difficult than the `unDefer` bussiness in the same file.

**Comment by serras** [comment]
It took a bit more than expected, mostly because some tests broke because of that. In fact, locally all the benchmarks fail for me, but I haven't been able to fix them.

File: src/Development/IDE/Core/Compile.hs [hunk 1]
```diff
@@ -247,10 +250,55 @@ upgradeWarningToError (nfp, sh, fd) =
   warn2err = T.intercalate ": error:" . T.splitOn ": warning:"
 
 hideDiag :: DynFlags -> (WarnReason, FileDiagnostic) -> (WarnReason, FileDiagnostic)
-hideDiag originalFlags (Reason warning, (nfp, _sh, fd))
-  | not (wopt warning originalFlags) = (Reason warning, (nfp, HideDiag, fd))
+hideDiag originalFlags (Reason warning, (nfp, sh, fd))
+  | not (wopt warning originalFlags)
+  = if warning `elem` unnecessaryDeprecationWarningFlags
+       then (Reason warning, (nfp, sh, fd{_severity = Just DsInfo}))
+       else (Reason warning, (nfp, HideDiag, fd))
```
Comment (ndmitchell) [inline src/Development/IDE/Core/Compile.hs:257]: I assume HideDiag means they don't appear at all? Could you just look at _tags ahd only Hide the ones with no tags? (I've no idea here, somewhat guessing)

File: src/Development/IDE/Core/Compile.hs [hunk 2]
```diff
@@ -247,10 +250,55 @@ upgradeWarningToError (nfp, sh, fd) =
...
+#if MIN_GHC_API_VERSION(8,10,0)
+    , Opt_WarnUnusedRecordWildcards
+#endif
+#if MIN_GHC_API_VERSION(8,6,0)
+    , Opt_WarnInaccessibleCode
+#endif
+    , Opt_WarnWarningsDeprecations
+    ]
+
+tagDiag :: (WarnReason, FileDiagnostic) -> (WarnReason, FileDiagnostic)
```
Comment (ndmitchell) [inline src/Development/IDE/Core/Compile.hs:284]: Could we get a comment on what this does and why? (From the code, its clear, but nice to be able to eyeball it without reading the code)

File: test/src/Development/IDE/Test.hs [hunk 3]
```diff
@@ -76,7 +83,11 @@ expectNoMoreDiagnostics timeout = do
     ignoreOthers = void anyMessage >> handleMessages
 
 expectDiagnostics :: [(FilePath, [(DiagnosticSeverity, Cursor, T.Text)])] -> Session ()
-expectDiagnostics expected = do
+expectDiagnostics
+  = expectDiagnostics_ . map (\(fp, dd) -> (fp, map (\(ds, c, t) -> (ds, c, t, Nothing)) dd))
+
+expectDiagnostics_ :: [(FilePath, [(DiagnosticSeverity, Cursor, T.Text, Maybe DiagnosticTag)])] -> Session ()
+expectDiagnostics_ expected = do
```
Comment (ndmitchell) [inline test/src/Development/IDE/Test.hs:90]: Ususally `foo_` would be the version of `foo` that does slightly less than `foo`, e.g. doesn't return some info. Could we call `expectDiagnostics_` something like `expectedDiagnosticsTags`?

File: test/exe/Main.hs [hunk 4]
```diff
@@ -3065,6 +3083,8 @@ ifaceErrorTest = testCase "iface-error-test-1" $ withoutStackEnv $ runWithExtraF
     --   - P is being typechecked with the last successful artifacts for A.
     expectDiagnostics [("P.hs", [(DsWarning,(4,0), "Top-level binding")])
                       ,("P.hs", [(DsWarning,(6,0), "Top-level binding")])
+                      ,("P.hs", [(DsInfo,(4,0), "Defined but not used")])
```
Comment (ndmitchell) [inline test/exe/Main.hs:3086]: The really important thing is that these diagnostics have the right tag associated with them. While they don't all need to be `expectDiagnostics_` (or whatever it ends up being called), a few more being that with the tag assertion would be good.

**Review by ndmitchell — APPROVED** [review]
Looks awesome!

File: src/Development/IDE/Core/Compile.hs (hunk 1 above)
Comment (serras) [reply src/Development/IDE/Core/Compile.hs]: I think that's a good idea: make those with tags "less important" but not hide them.

Implemented in the latest commit.

File: src/Development/IDE/Core/Compile.hs (hunk 2 above)
Comment (serras) [reply src/Development/IDE/Core/Compile.hs:286]: Added comment to the file:

> -- | Add a unnecessary/deprecated tag to the required diagnostics.

File: test/src/Development/IDE/Test.hs (hunk 3 above)
Comment (serras) [reply test/src/Development/IDE/Test.hs]: I've renamed it to `expectDiagnosticsWithTags`.

File: test/exe/Main.hs (hunk 4 above)
Comment (serras) [reply test/exe/Main.hs]: That makes sense. I've extended the checks in the latest commit.

**Comment by ndmitchell** [comment]
I see some test failure:

```
FAIL (0.59s)
      test/src/Development/IDE/Test.hs:77:
      Got unexpected diagnostics for Uri {getUri = "file:///tmp/extra-dir-28584335661024/ModuleA.hs"} got List [Diagnostic {_range = Range {_start = Position {_line = 1, _character = 0}, _end = Position {_line = 1, _character = 29}}, _severity = Just DsInfo, _code = Nothing, _source = Just "typecheck", _message = "The import of \8216ModuleB\8217 is redundant\n  except perhaps to import instances from \8216ModuleB\8217\nTo import instances alone, use: import ModuleB()", _tags = Just (List [DtUnnecessary]), _relatedInformation = Nothing}]
```

I did have to fix a merge conflict to get it going again. So certainly possible that's my fault.

**Comment by serras** [comment]
> I see some test failure:
> 
> ```
> FAIL (0.59s)
>       test/src/Development/IDE/Test.hs:77:
>       Got unexpected diagnostics for Uri {getUri = "file:///tmp/extra-dir-28584335661024/ModuleA.hs"} got List [Diagnostic {_range = Range {_start = Position {_line = 1, _character = 0}, _end = Position {_line = 1, _character = 29}}, _severity = Just DsInfo, _code = Nothing, _source = Just "typecheck", _message = "The import of \8216ModuleB\8217 is redundant\n  except perhaps to import instances from \8216ModuleB\8217\nTo import instances alone, use: import ModuleB()", _tags = Just (List [DtUnnecessary]), _relatedInformation = Nothing}]
> ```
> 
> I did have to fix a merge conflict to get it going again. So certainly possible that's my fault.

I've spent some time and fixed the test. Hopefully everything should be green in some minutes.

**Comment by ndmitchell** [comment]
Awesome work - this is a great feature. Thanks for all your hard to work to write it and adjust so much of the test suite.

**Pull request merged**